src/plugins/content_management/server/core/msearch.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { SavedObjectsClientContract } from '@kbn/core-saved-objects-api-server';
+import type { MSearchResult, SearchQuery } from '../../common';
+import { ContentRegistry } from './registry';
+import { StorageContext } from './types';
+
+export class MSearchService {
+  constructor(
+    private readonly deps: {
+      getSavedObjectsClient: () => Promise<SavedObjectsClientContract>;
+      contentRegistry: ContentRegistry;
+    }
+  ) {}
+
+  async search(
```

Comment:
Looks good. I had to put a few comments and spaces to make it slightly more readable (quite a few `Map` around!).

That's how I left it in local

```ts
async search(
    contentTypes: Array<{ contentTypeId: string; ctx: StorageContext }>,
    query: SearchQuery
  ): Promise<MSearchResult> {
    // Map: contentTypeId -> StorageContext
    const contentTypeToCtx = new Map(contentTypes.map((ct) => [ct.contentTypeId, ct.ctx]));

    // Map: contentTypeId -> MSearchConfig
    const contentTypeToMSearchConfig = new Map(
      contentTypes.map((ct) => {
        const mSearchConfig = this.deps.contentRegistry.getDefinition(ct.contentTypeId).storage
          .mSearch;
        if (!mSearchConfig) {
          throw new Error(`Content type ${ct.contentTypeId} does not support mSearch`);
        }
        return [ct.contentTypeId, mSearchConfig];
      })
    );

    // Map: Saved object type -> [contentTypeId, MSearchConfig]
    const soTypeToMSearchConfig = new Map(
      Array.from(contentTypeToMSearchConfig.entries()).map(([ct, mSearchConfig]) => {
        return [mSearchConfig.savedObjectType, [ct, mSearchConfig] as const];
      })
    );

    const mSearchConfigs = Array.from(contentTypeToMSearchConfig.values());
    const soSearchTypes = mSearchConfigs.map((mSearchConfig) => mSearchConfig.savedObjectType);

    const additionalSearchFields = new Set<string>();
    mSearchConfigs.forEach((mSearchConfig) => {
      (mSearchConfig.additionalSearchFields ?? []).forEach(additionalSearchFields.add);
    });

    const savedObjectsClient = await this.deps.getSavedObjectsClient();
    const soResult = await savedObjectsClient.find({
      type: soSearchTypes,
      search: query.text,
      searchFields: [`title^3`, `description`, ...additionalSearchFields],
      defaultSearchOperator: 'AND',
      // TODO: tags
      // TODO: pagination
      // TODO: sort
    });
   
   ...
```